components/concept/storage/src/main/java/mozilla/components/concept/storage/BookmarksStorage.kt

```diff
@@ -0,0 +1,116 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.concept.storage
+
+interface BookmarksStorage {
+
+    /**
+     * Produces a bookmarks tree for the given guid string.
+     *
+     * @param guid The bookmark guid to obtain.
+     * @param recursive Whether to recurse and obtain all levels of children.
+     * @return The populated root starting from the guid.
+     */
+    suspend fun getTree(guid: String, recursive: Boolean = false): BookmarkNode?
+
+    /**
+     * Searches bookmarks with a query string.
+     *
+     * @param query The query string to search.
+     * @param limit The maximum number of entries to return.
+     * @return The list of matching bookmark nodes up to the limit number of items.
+     */
+    suspend fun searchBookmarks(query: String, limit: Int = defaultBookmarkSearchLimit): List<BookmarkNode>
+
+    /**
+     * Adds a new bookmark item to a given node.
+     *
+     * Sync behavior: will add new bookmark item to remote devices.
+     *
+     * @param parentGuid The parent guid of the new node.
+     * @param newNode The details of the node to add.
+     * @param position The optional position to add the new node or null to append.
+     * @return The guid of the newly inserted bookmark item.
+     */
+    suspend fun addItem(parentGuid: String, newNode: BookmarkNode, position: Int? = null): String
+
+    /**
+     * Adds a new bookmark folder to a given node.
+     *
+     * Sync behavior: will add new separator to remote devices.
+     *
+     * @param parentGuid The parent guid of the new node.
+     * @param newNode The details of the node to add.
+     * @param position The optional position to add the new node or null to append.
+     * @return The guid of the newly inserted bookmark item.
+     */
+    suspend fun addFolder(parentGuid: String, newNode: BookmarkNode, position: Int? = null): String
```

Comment:
You are using a `BookmarkNode` when a `title` would have been sufficient. `BookmarkNode` as you defined it could be a bookmark, a folder or a separator, and it will always have a `guid`. It might even have a `url`. None of those are valid inputs for this API, only `title` and `position` are.

Another way to think about this - `add*` APIs should take a generic description of an item (where, what), not a reified version of it (an object with an assigned `guid`). Doing it the other way is akin to creating a variable and insisting on a memory address, as opposed to asking the memory allocator to give you one for the data you're trying to store.

If you really want to have this API take an object, you'll need to define a `PendingBookmarkFolder`, or a `BookmarkFolder` with an optional `guid`... These options seem worse to me.

Reply:
Fixed